code/modules/research/machinery/part_fabricator/part_fabricator.dm

```diff
@@ -0,0 +1,521 @@
+			if(2)
+				production_speed *= 0.5
+			if(3)
+				production_speed *= 0.75
+			if(5 to INFINITY)
+				production_speed += 0.2 // at 21 parts, this provides a maximum speed boost of 5.2x
+	if(reagents)
+		reagents.maximum_volume = 0
+		for(var/obj/item/reagent_containers/glass/G in component_parts)
+			reagents.maximum_volume += G.volume
+			G.reagents.trans_to(src, G.reagents.total_volume)
+
+/obj/machinery/part_fabricator/on_deconstruction()
+	for(var/obj/item/reagent_containers/glass/G in component_parts)
+		reagents.trans_to(G, G.reagents.maximum_volume)
+	for(var/obj/item/item in contents)
+		item.forceMove(get_turf(src)) // Eject anything we may be holding
+		adjust_item_drop_location(item)
+
+/obj/machinery/part_fabricator/Initialize()
+	. = ..()
+	create_reagents(0, OPENCONTAINER)
+	RefreshParts()
+
+	var/list/occupied = list()
+	for(var/direct in list(EAST,WEST,NORTH))
+		occupied += get_step(src,direct)
+
+	for(var/T in occupied)
+		var/obj/structure/filler/F = new(T)
+		F.parent = src
+		fillers += F
+
+	if(part_recipes_generated)
+		return
+	
+	capacitor_energy_requirement = (rand() * 0.5 + 0.75) * 1000000000 // 0.75-1.25 GW
+
+	matterbin_freon_moles_requirement = (rand() * 0.5 + 0.5) * 100 // 50-100 moles
+
+	scanner_chemicals_requirement = list()
+	scanner_chemicals_requirement += new /datum/bounty/reagent/simple_drink
+	scanner_chemicals_requirement += new /datum/bounty/reagent/chemical_simple
+
+	laser_money_requirement = round((rand() * 0.5 + 0.75) * 10000) // 7500-12500 credits
+
+	var/list/possible_plants = subtypesof(/datum/bounty/item/botany)
+	for(var/datum/bounty/item/botany/plant_bounty in possible_plants)
+		if(initial(plant_bounty.multiplier) == 0)
+			possible_plants -= plant_bounty
+	var/bounty_type = pick(possible_plants)
+	manipulator_plant_requirement = new bounty_type
+
+	manipulator_temp_requirement = (rand() + 1) * 40000 // 40000-80000 Kelvin
+
+	acceptable_items = typecacheof(list(
+		/obj/item/electrical_stasis_manifold,
+		/obj/item/organic_augur,
+		/obj/item/mmi/posibrain,
+		/obj/item/gun/energy/laser,
+		/obj/item/organ
+	))
+
+	for(var/item in manipulator_plant_requirement.wanted_types)
+		acceptable_items += typecacheof(item)
+
+	part_recipes_generated = TRUE
+
+/obj/machinery/part_fabricator/Destroy()
+	for(var/V in fillers)
+		var/obj/structure/filler/filler = V
+		filler.parent = null
+		qdel(filler)
+	. = ..()
+
+/obj/machinery/part_fabricator/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!is_operational())
+		return
+	if(!ui)
+		ui = new(user, src, "PartFabricator", name)
+		ui.open()
+
+/obj/machinery/part_fabricator/ui_data(mob/user)
+	var/list/data = ..()
+	// Capacitor requirements /////////////////////////////////////////////////////////////////
+	if(tab == "capacitor")
+		var/current_ESMs = 0
+		for(var/obj/item/electrical_stasis_manifold/esm in contents)
+			current_ESMs++
+		data["current_ESMs"] = current_ESMs ? current_ESMs : "0"
```

Comment:
```suggestion
		data["current_ESMs"] = current_ESMs ? current_ESMs : 0
```
for the rest too, pretty sure it won't drop the data

Reply:
pretty sure it does